In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
path=r"C:\Users\devis\OneDrive\Documents\data science\data files\Visadataset.csv"
visa_df=pd.read_csv(path)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25480 rows x 12 columns]

In [5]:
cat_columns=visa_df.select_dtypes(include='object').columns
num_columns=visa_df.select_dtypes(exclude='object').columns
num_columns

Index(['no_of_employees', 'yr_of_estab', 'prevailing_wage'], dtype='object')

**Encoding**
- encoding means convert cat columns data to numerical data 
- beacause ML models expect the data should be in numerical format
- ML models developed by math algorithms
- it is very very important to apply encoding techniques
- the following methods are
   - map
   - np.where
   - one hot encoding
   - label encoder

**map method**

- mapping is one method to convert categorical values to numerical
- take one categorical column
- get the unique label first
- make a dictionary with assign  a number to each label
- for example case_status has 2 labels
  - certified
  - denied
- assign 0 to certified and  1 to denied
- create a dictionary labels as keys and numbers as values only
- {'certified':0,'denied':1}


In [6]:
visa_df['case_status'].unique()

array(['Denied', 'Certified'], dtype=object)

In [7]:

d={'Certified':0,'Denied':1}
visa_df['case_status_new']=visa_df['case_status'].map(d)
visa_df[['case_status_new']]


case_status_new
0                    1
1                    0
2                    1
3                    1
4                    0
...                ...
25475                0
25476                0
25477                0
25478                0
25479                0

[25480 rows x 1 columns]

In [8]:
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  case_status_new  
0          Denied                1  
1       Certified                0  
2          Denied                1  
3          Denied                1  
4       Certified                0  
...           ...              ...  
25475   Certified                0  
25476   Certified                0  
25477   Certified                0  
25478   Certified                0  
25479   Certified                0  

[25480 rows x 13 columns]

In [9]:
visa_df['continent'].unique()

array(['Asia', 'Africa', 'North America', 'Europe', 'South America',
       'Oceania'], dtype=object)

In [10]:
d={'Asia':0, 'Africa':1, 'North America':2, 'Europe':3, 'South America':4,
       'Oceania':5}
visa_df['continent_new']=visa_df['continent'].map(d)
visa_df[['continent_new']]


continent_new
0                  0
1                  0
2                  0
3                  0
4                  1
...              ...
25475              0
25476              0
25477              0
25478              0
25479              0

[25480 rows x 1 columns]

In [11]:
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  case_status_new  continent_new  
0          Denied                1              0  
1       Certified                0              0  
2          Denied                1              0  
3          Denied                1              0  
4       Certified                0              1  
...           ...              ...            ...  
25475   Certified                0              0  
25476   Certified                0              0  
25477   Certified                0              0  
25478   Certified                0              0  
25479   Certified                0              0  

[25480 rows x 14 columns]

In [17]:
visa_df['education_of_employee'].unique()

array(['High School', "Master's", "Bachelor's", 'Doctorate'], dtype=object)

In [12]:
d2={'High School':0, "Master's":1, "Bachelor's":2, 'Doctorate':3}
visa_df['education_of_employee_new']=visa_df['education_of_employee'].map(d2)
visa_df[['education_of_employee_new']]


education_of_employee_new
0                              0
1                              1
2                              2
3                              2
4                              1
...                          ...
25475                          2
25476                          0
25477                          1
25478                          1
25479                          2

[25480 rows x 1 columns]

In [13]:
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  case_status_new  continent_new  education_of_employee_new  
0          Denied                1              0                          0  
1       Certified                0              0                          1  
2          Denied                1              0                          2  
3          Denied                1              0                          2  
4       Certified                0              1                          1  
...           ...              ...            ...                        ...  
25475   Certified                0              0                          2  
25476   Certified                0              0                          0  
25477   Certified                0              0                          1  
25478   Certified                0              0                          1  
25479   Certified                0              0                          2  

[25480 rows x 15 columns]

In [14]:
visa_df['region_of_employment'].unique()

array(['West', 'Northeast', 'South', 'Midwest', 'Island'], dtype=object)

In [23]:
d3={'West':100, 'Northeast':101, 'South':102, 'Midwest':103, 'Island':104}
visa_df['region_of_employe_new']=visa_df['region_of_employment'].map(d3)
visa_df[['region_of_employe_new']]


region_of_employe_new
0                        100
1                        101
2                        100
3                        100
4                        102
...                      ...
25475                    102
25476                    101
25477                    102
25478                    100
25479                    103

[25480 rows x 1 columns]

In [15]:
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  case_status_new  continent_new  education_of_employee_new  
0          Denied                1              0                          0  
1       Certified                0              0                          1  
2          Denied                1              0                          2  
3          Denied                1              0                          2  
4       Certified                0              1                          1  
...           ...              ...            ...                        ...  
25475   Certified                0              0                          2  
25476   Certified                0              0                          0  
25477   Certified                0              0                          1  
25478   Certified                0              0                          1  
25479   Certified                0              0                          2  

[25480 rows x 15 columns]

In [ ]:
#ATS

In [21]:
unique_labels=(visa_df['case_status'].unique())
list1=[i for i in range(len(unique_labels))]
dict1=dict(zip(unique_labels,list1))
dict1

{'Denied': 0, 'Certified': 1}

In [22]:
{key:value for key,value in zip(unique_labels,list1)}

{'Denied': 0, 'Certified': 1}

In [25]:
unique_labels=(visa_df['case_status'].unique())
list1=[i for i in range(len(unique_labels))]
dict1=dict(zip(unique_labels,list1))
dict1
visa_df['case_status_df']=visa_df['case_status'].map(dict1)
visa_df[['case_status_df']]

case_status_df
0                   0
1                   1
2                   0
3                   0
4                   1
...               ...
25475               1
25476               1
25477               1
25478               1
25479               1

[25480 rows x 1 columns]

In [26]:
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  case_status_new  continent_new  education_of_employee_new  \
0          Denied                1              0                          0   
1       Certified                0              0                          1   
2          Denied                1              0                          2   
3          Denied                1              0                          2   
4       Certified                0              1                          1   
...           ...              ...            ...                        ...   
25475   Certified                0              0                          2   
25476   Certified                0              0                          0   
25477   Certified                0              0                          1   
25478   Certified                0              0                          1   
25479   Certified                0              0                          2   

       case_status_df  
0                   0  
1                   1  
2                   0  
3                   0  
4                   1  
...               ...  
25475               1  
25476               1  
25477               1  
25478               1  
25479               1  

[25480 rows x 16 columns]

In [34]:
visa_df=pd.read_csv(r"C:\Users\devis\OneDrive\Documents\data science\data files\Visadataset.csv")
cat_columns=visa_df.select_dtypes(include='object').columns
for i in cat_columns:
    unique_labels=sorted(visa_df[i].unique())
    list1=[i for i in range(len(unique_labels))]
    dict1=dict(zip(unique_labels,list1))
    dict1
    visa_df[i]=visa_df[i].map(dict1)
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

**label encoder**
- labelencoder is a method to convert categorical data to numerical data
- labelencoder is a same way how we did in the above
- the for loop map method we explored
- but the label encoder also do the same
- it is under **sickit_learn** we usually called as **sklearn**
- under sk learn we have calss called as preprossing
- because currently we are doing data preprocessing only
- under preprocessing class we have method called as **lablencoder**
- sklearn
    - preprocessing
    - lablencoder

- any sklearn package has 3 steps
  
  - step-1:read the package
  - step-2:save the package
  - step-3:apply fit transform on data

In [36]:
#read the data
visa_df=pd.read_csv(r"C:\Users\devis\OneDrive\Documents\data science\data files\Visadataset.csv")
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25480 rows x 12 columns]

In [37]:
# step-1:read the package/load the method
from sklearn.preprocessing import LabelEncoder

# step-2:save the package/call the method
le=LabelEncoder()

#step-3:apply fit transform on data
visa_df['case_status']=le.fit_transform(visa_df['case_status'])
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                0  
2                1  
3                1  
4                0  
...            ...  
25475            0  
25476            0  
25477            0  
25478            0  
25479            0  

[25480 rows x 12 columns]

In [38]:
visa_df=pd.read_csv(r"C:\Users\devis\OneDrive\Documents\data science\data files\Visadataset.csv")

In [39]:
cat_columns=visa_df.select_dtypes(include='object').columns
for i in cat_columns:
    from sklearn.preprocessing import LabelEncoder
    le=LabelEncoder()
    visa_df[i]=le.fit_transform(visa_df[i])
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

In [40]:
visa_df=pd.read_csv(r"C:\Users\devis\OneDrive\Documents\data science\data files\Visadataset.csv")

In [41]:
from sklearn.preprocessing import LabelEncoder
cat_columns=visa_df.select_dtypes(include='object').columns
le=LabelEncoder()
for i in cat_columns:
    visa_df[i]=le.fit_transform(visa_df[i])
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

# fit and transform

- fit means develop the logic
- transform means process the logic
- if we observe in map first we develop dictionary logic
- then we apply the dictionary to column
- developing the dictionary is kind of fit
- apply the dictionary to column means we are transforming the data cat to num cols
- whenever we are transforming or changing the data then use **fit transform**
- if we are not changing the data then only **fit**
    

**np.where**
- using np.where also we can change the categorical data to numerical 
- but np.where is only for binary condition
- np.where is same as if and else
- if condition is true then we apply true value
- if the condition is false we apply false value
- so np.where is use for only binary labels
- a column should have only 2 unique labels

In [42]:
# value=='certified'=====  replace with 0===otherwise 1
visa_df=pd.read_csv(r"C:\Users\devis\OneDrive\Documents\data science\data files\Visadataset.csv")

In [48]:
con=visa_df['case_status']=='Certified'
visa_df['case_status']=np.where(con,0,1)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                0  
2                1  
3                1  
4                0  
...            ...  
25475            0  
25476            0  
25477            0  
25478            0  
25479            0  

[25480 rows x 12 columns]

**one hot encoder**

- one hot means one will be ON, anther will be OFF
- ON reprsents with 1
- OFF represents with 0
- for example case status has 2 unique labels 
  - certified
  - denied
- one hot encoder creates new column which is equal to no of unique labels
- for example case_status will create two new extra columns
    - case_status_certified
    - case_status_denied

- case_status     case_status_certified      case_status_denied

  denied                0                           1
  certified             1                           0
  denied                0                           1

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
path=r"C:\Users\devis\OneDrive\Documents\data science\data files\Visadataset.csv"
visa_df=pd.read_csv(path)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25480 rows x 12 columns]

In [3]:
cat_columns=visa_df.select_dtypes(include='object').columns
num_columns=visa_df.select_dtypes(exclude='object').columns
num_columns

Index(['no_of_employees', 'yr_of_estab', 'prevailing_wage'], dtype='object')

**pd.get_dummies**

In [4]:
pd.get_dummies(visa_df['case_status'])

Certified  Denied
0          False    True
1           True   False
2          False    True
3          False    True
4           True   False
...          ...     ...
25475       True   False
25476       True   False
25477       True   False
25478       True   False
25479       True   False

[25480 rows x 2 columns]

In [5]:
pd.get_dummies(visa_df['case_status'],dtype='int')

Certified  Denied
0              0       1
1              1       0
2              0       1
3              0       1
4              1       0
...          ...     ...
25475          1       0
25476          1       0
25477          1       0
25478          1       0
25479          1       0

[25480 rows x 2 columns]

In [6]:
df=pd.get_dummies(visa_df['case_status'],dtype='int')
df['case_status']=visa_df['case_status'].values
df

Certified  Denied case_status
0              0       1      Denied
1              1       0   Certified
2              0       1      Denied
3              0       1      Denied
4              1       0   Certified
...          ...     ...         ...
25475          1       0   Certified
25476          1       0   Certified
25477          1       0   Certified
25478          1       0   Certified
25479          1       0   Certified

[25480 rows x 3 columns]

In [7]:
df=pd.get_dummies(visa_df['case_status'],dtype='int')
df['case_status']=visa_df['case_status'].values
df['dummy']=[i for i in range(len(visa_df))]
df

Certified  Denied case_status  dummy
0              0       1      Denied      0
1              1       0   Certified      1
2              0       1      Denied      2
3              0       1      Denied      3
4              1       0   Certified      4
...          ...     ...         ...    ...
25475          1       0   Certified  25475
25476          1       0   Certified  25476
25477          1       0   Certified  25477
25478          1       0   Certified  25478
25479          1       0   Certified  25479

[25480 rows x 4 columns]

In [15]:
#df1=pd.get_dummies(visa_df['case_status'],dtype='int')
#df2=visa_df[['case_status']]
#pd.concat(df1,df2,axis=1)

# adding 2 dataframes
- merge
- join
- concat
- append

In [17]:
pd.get_dummies(visa_df,dtype='int')

no_of_employees  yr_of_estab  prevailing_wage  case_id_EZYV01  \
0                14513         2007         592.2029               1   
1                 2412         2002       83425.6500               0   
2                44444         2008      122996.8600               0   
3                   98         1897       83434.0300               0   
4                 1082         2005      149907.3900               0   
...                ...          ...              ...             ...   
25475             2601         2008       77092.5700               0   
25476             3274         2006      279174.7900               0   
25477             1121         1910      146298.8500               0   
25478             1918         1887       86154.7700               0   
25479             3195         1960       70876.9100               0   

       case_id_EZYV02  case_id_EZYV03  case_id_EZYV04  case_id_EZYV05  \
0                   0               0               0               0   
1                   1               0               0               0   
2                   0               1               0               0   
3                   0               0               1               0   
4                   0               0               0               1   
...               ...             ...             ...             ...   
25475               0               0               0               0   
25476               0               0               0               0   
25477               0               0               0               0   
25478               0               0               0               0   
25479               0               0               0               0   

       case_id_EZYV06  case_id_EZYV07  ...  region_of_employment_South  \
0                   0               0  ...                           0   
1                   0               0  ...                           0   
2                   0               0  ...                           0   
3                   0               0  ...                           0   
4                   0               0  ...                           1   
...               ...             ...  ...                         ...   
25475               0               0  ...                           1   
25476               0               0  ...                           0   
25477               0               0  ...                           1   
25478               0               0  ...                           0   
25479               0               0  ...                           0   

       region_of_employment_West  unit_of_wage_Hour  unit_of_wage_Month  \
0                              1                  1                   0   
1                              0                  0                   0   
2                              1                  0                   0   
3                              1                  0                   0   
4                              0                  0                   0   
...                          ...                ...                 ...   
25475                          0                  0                   0   
25476                          0                  0                   0   
25477                          0                  0                   0   
25478                          1                  0                   0   
25479                          0                  0                   0   

       unit_of_wage_Week  unit_of_wage_Year  full_time_position_N  \
0                      0                  0                     0   
1                      0                  1                     0   
2                      0                  1                     0   
3                      0                  1                     0   
4                      0                  1                     0   
...                  ...                ...                  

- in one hot encoder a new column will be created
- the new columns equal to no of unique labels
- because of case status to extra columns are created
- because of continent 6 extra columns arre created
- because of case_id 25480 unique ids,25480 columns are created

In [23]:
count=0
for i in cat_columns:
    count=count+visa_df[i].nunique()

print(count+len(num_columns))

25510


In [24]:
count=0
for i in cat_columns[1:]:
    count=count+visa_df[i].nunique()
print(count)
print(count+len(num_columns))

27
30


In [25]:
for i in cat_columns[1:]:
    print(visa_df[i].unique(),visa_df[i].nunique())

['Asia' 'Africa' 'North America' 'Europe' 'South America' 'Oceania'] 6
['High School' "Master's" "Bachelor's" 'Doctorate'] 4
['N' 'Y'] 2
['N' 'Y'] 2
['West' 'Northeast' 'South' 'Midwest' 'Island'] 5
['Hour' 'Year' 'Week' 'Month'] 4
['Y' 'N'] 2
['Denied' 'Certified'] 2


**advantages**
- using one hot encoder we will ge new columns
- the new column values at a time only one will be **True** another will be **False**
- which means new columns are **independent each other**
- independent means **no relation**
- means no **multi collineraity**
    -multi collineraity is an imp property in ML
- **machine learning models expect the columns should be independent each other**
- so one hot encoder gives the no multicollineraity
    


**disadvantages**
  
- the disadv is it will increase many columns
- in the above example the total cols becomes 25510
- this is happen because of case id
- in real time we will drop the id columns
- eventhough if we drop still we are getting 30 cols
- more columns means more time to process
- we required more infrastructure(CPU-GPU-TPU)
- this is called as **curse od dimensionality** 

In [13]:
df1=pd.get_dummies(visa_df['case_status'],dtype='int')
df2=visa_df[['case_status']]


In [14]:
df11=pd.concat([df1,df2],axis=1)
df11

Certified  Denied case_status
0              0       1      Denied
1              1       0   Certified
2              0       1      Denied
3              0       1      Denied
4              1       0   Certified
...          ...     ...         ...
25475          1       0   Certified
25476          1       0   Certified
25477          1       0   Certified
25478          1       0   Certified
25479          1       0   Certified

[25480 rows x 3 columns]

In [20]:
merge_df=pd.merge(df1,df2)
merge_df

MergeError: No common columns to perform merge on. Merge options: left_on=None, right_on=None, left_index=False, right_index=False